### ISSUE TITLE
Gitea markdown not sanitised (links can contain API URLs)

### ISSUE THREAD

**Issue by Siesh1oo** [go-gitea/gitea#4596]
- Gitea version (or commit ref): 1.5.0+rc1-94-g819f50ccd
- Git version: nevermind
- Operating system: nevermind
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io: probably (currently down "bad gateway")
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description

Create README.md or other markdown file containing a link to some gitea API endpoint:

```
[I might be a misleading, yet harmless looking link. Click me to learn more](/user/logout)
```

## Problem

API endpoints are normal GET requests, not protected by auth (normal session id used).

Markdown is rendered in normal (non-sandboxed) DOM.

## Potential solutions

1. sanitize URLs (do not allow relative or absolute links to API endpoints; if so, regex search'n'replace can for example fill with zero-width/invisible unicode whitespace, or just strike it through),
2. sandbox markdown rendered DOM in iframe with sandbox attribute set (and remove session cookie from outside using JavaScript),
3. Enforce CSRF for all API endpoints.

ideally all three

**Comment by Siesh1oo** [comment]
## Addendum

Same bug for images. Note that the user does not even has to click the image, loading the page will execute the injected API code.

Testcase: create README.md containing an image with URL to API endpoint, log in, visit repository page, do something else and realise that gitea has logged you out:

```
![This is not an Image of Yaktocat](/user/logout)
```

**Comment by cezar97** [comment]
Hello,

> Enforce CSRF for all API endpoints.

Same as #4357.

> API endpoints are normal GET requests.

> Same bug for images. Note that the user does not even has to click the image, loading the page will execute the injected API code.

Same as #4337.
I complicated myself in trying to add this image that send a GET request in the repository's description. I should added it in the Markdown like you did :smile:.

**Comment by lafriks** [comment]
Yes this should be cleaned up but no harming actions can be done as GET requests are only for getting data not doing actions (except for logout url)

**Comment by Siesh1oo** [comment]
@lafriks: Are you sure? The team/org router go suggested otherwise; logout was used above only as harmless example

**Comment by Siesh1oo** [comment]
Maybe some collaboration possible: https://github.com/gogs/gogs/issues/5355#issuecomment-410117204

**Comment by Siesh1oo** [comment]
@cezar97:

> I complicated myself in trying to add this image that send a GET request in the repository's description. I should added it in the Markdown like you did 😄.

The two issues seem distinct, would need distinct fixes, if solved via sanitizing as discussed there:  removing/sanitizing HTML tags from markdown wouldn't fix this one. Markdown links need special treatment unless URLs and images are completely forbidden, which is probably not desirable

**Comment by cezar97** [comment]
@Siesh1oo:
> The two issues seem distinct, would need distinct fixes...

Yes, but the point was that logout and others are GET requests. The fact that I can send this requests from images in Gitea makes all more serious, but for this wasn't needed as I can add the same images on my site.
BTW inserting the images in the descriptions means that you don't have to click on the repository to open the README in markdown to send this GET, you just have to see my profile or the Explore Tab. My Proof of Concept was that after some weeks, I have more than 200 followers on try.gitea.io, as explained in #4337. So @lafriks, except for logout url **and** user followers **and** watchers of a repository **and** stars of a repository :) .

**Comment by msg7086** [comment]
I'd say, actions that can change system states, such as logout or star, shall be POST only.

**Comment by cezar97** [comment]
POST can also be automated with a form. I write about it in #4357.

**Comment by msg7086** [comment]
Thanks for pointing out. Yes CSRF is also needed in this particular case. Just that my reply was focusing on best practices about HTTP verbs (e.g. RESTful style) and not about XSS protection.

Cheers

**Comment by Siesh1oo** [comment]
One more to check, this one had affected gitlab too: 
> server-side request forgery (SSRF) vulnerability in migrate https://github.com/gogs/gogs/issues/5372

**Comment by cezar97** [comment]
> Just that my reply was focusing on best practices...

Oh, then I agree with you. But adding a CSRF check too it's better indeed.

**Comment by jolheiser** [comment]
Fixed by #10462, #10465, #10582
